common/hooks/post-install: add fix permissions hook

paper42 opened the pull request:
Some packages install files with wrong permissions, but sometimes we can detect and fix them.

## /usr/share/man: 644
this rule matches a lot of files, mainly because it matches 444 permissions too
* packages which install manpages with 755 permissions: nvimpager, sloccount
* packages which install manpages with 444 permissions: lowdown, mdocml, dhcpcd, openresolv, all perl packages, lua5.3 (but not 5.1, 5.2 and 5.4)

## /etc/apparmor.d: 600
I chose 600 because that's what aa-genprof creates.
* packages which install apparmor profiles wrong permissions: apparmor (644), brillo (640), firejail (644), mako (640) (these permission measurements may be wrong in some cases)

The package lists are not complete.

Are there any other common directories which should be included in this hook? Is forcing 644 in /usr/share/man too strict? Should affected packages be revbumped?

Duncaen commented on common/hooks/post-install/14-fix-permissions.sh:
```diff
@@ -0,0 +1,19 @@
+	dir="$PKGDESTDIR$1"
+	perms="$2"
+	if [ -d "$dir" ]; then
+		find "$dir" -type f | while read -r file; do
```
find "$dir" -type f ! -perm -0644

Duncaen replied:
Or even better:
```find /usr/include/ -type f ! -perm -"$perms" -exec chown "$perms" '{}' +```

paper42 replied:
`! -perm -0600` would match a file with permissions 700, but not a file with permissions 670. I modified the function to accept a permission mask as its second argument.

> Or even better:
> find /usr/include/ -type f ! -perm -"$perms" -exec chmod -v "$perms" '{}' +

Is this better? It's shorter, but also we lose the pretty warning message.

Duncaen replied:
Shorter and faster and `chmod -v` still gives you messages.

paper42 replied:
Thank you, fixed.

Duncaen commented:
Not sure of forcing 600 on /etc/apparmor.d is a good idea, I don't see no harm in them being readable.

Would also make sense to also make sure nothing is o+w, there were/are a few packages that did that:
https://gist.github.com/Duncaen/125a44a4e9f159141bcaade111a182e6
In those cases its probably better to abort instead of trying to fix them.

ericonr commented:
I think revbumping is warranted if any packages are writable when they shouldn't be, or if permissions are a security concern for that case.

I agree that apparmor files can be world readable - makes inspection easier. Do you have any argument for it beyond being the default from genprof?

paper42 commented:
> Not sure of forcing 600 on /etc/apparmor.d is a good idea, I don't see no harm in them being readable.

> I agree that apparmor files can be world readable - makes inspection easier. Do you have any argument for it beyond being the default from genprof?

I don't have a good reason, I just thought it was annoying having some profiles readable by my user and some not. I fixed the script to change permissions to 644 only if the profile is executable (mask 111).

> Would also make sense to also make sure nothing is o+w, there were/are a few packages that did that:
> https://gist.github.com/Duncaen/125a44a4e9f159141bcaade111a182e6
> In those cases its probably better to abort instead of trying to fix them.

Great idea, I added it to the script, it requires iterating over all files and in the script, so some directories are scanned twice, but I think it's not a big problem.

I also added 3 new rules:
* /usr/include - 644 - http-parser-devel (755), cups (444)

are there reasons to not force permissions here I don't know about?

* /usr/share/applications 644 - KDE applications often violate this rule - falkon, kate5, kde-cli-tools, kdevelop, khelpcenter, kinfocenter, kio, kmenuedit, knewstuff, konsole, kronometer, ksysguard, okteta, plasma-desktop, plasma-workspace, syncthing, systemsettings

* /usr/icons - 644 - no known violations

since there are no known violations, is it worth including this rule?

ericonr commented:
> since there are no known violations, is it worth including this rule?

I'd say yes, because it allows us to impose policy...

paper42 commented:
added some new rules, I don't know about any violations

* /usr/share/help 644
* /usr/share/icons" 644
* /usr/share/locale 644
* /usr/share/metainfo 644

paper42 commented:
@Duncaen @ericonr Is there anything you would like to add/change? It would be nice to get this merged.

Duncaen commented:
Seems fine to me, I wonder if there could be a case where we would need to allow a file to be writable by all users.
There are some `*.hscr` files that are at the moment, not sure if this should be allowed.

ericonr commented on common/hooks/post-install/14-fix-permissions.sh:
```diff
@@ -0,0 +1,28 @@
+	permmask="$2"
+	# permissions which will be set on matched files
+	perms="$3"
+	if [ -d "$dir" ]; then
+        find "$dir" -type f -perm /"$permmask" -exec chmod -v "$perms" {} +
+    fi
```
wonky indentation.

ericonr commented on common/hooks/post-install/14-fix-permissions.sh:
```diff
@@ -0,0 +1,28 @@
+# This hook fixes permissions in common places
+
+change_file_perms() {
+	dir="$PKGDESTDIR$1"
+	# permission mask for matching the files
+	permmask="$2"
+	# permissions which will be set on matched files
+	perms="$3"
```
I think it's standard to make these local variables.

ericonr commented on common/hooks/post-install/14-fix-permissions.sh:
```diff
@@ -0,0 +1,28 @@
+# This hook fixes permissions in common places
+
+change_file_perms() {
+	dir="$PKGDESTDIR$1"
```
```suggestion
	dir="${PKGDESTDIR}${1}"
```

To make reading easier, since it's appending.

ericonr commented on common/hooks/post-install/14-fix-permissions.sh:
```diff
@@ -0,0 +1,28 @@
+hook() {
+	# check that no files have permission write for all users
+	find "$PKGDESTDIR" -type f -perm -0002 | while read -r file; do
+		msg_error "$pkgver: file ${file#$PKGDESTDIR} has a write permission for all users\n"
```
```suggestion
		msg_error "$pkgver: file ${file#$PKGDESTDIR} has write permission for all users\n"
```

I think?

paper42 replied:
You are right, thanks!

paper42 commented:
> Seems fine to me, I wonder if there could be a case where we would need to allow a file to be writable by all users.
> There are some `*.hscr` files that are at the moment where it maybe makes sense, but I'm not sure if this should be allowed.

I added a new variable nofixperms which skips the whole check for the package. I will test how the packages which use .hscr files behave with this hook and fix them if necessary.

* added a rule for /usr/share/appdata which is a deprecated directory similar to /usr/share/metainfo

paper42 commented:
Lint is failing with `srcpkgs/lbreakout2/template:15: custom variables should use _ prefix: nofixperms=yes`, that's because xlint doesn't know about nofixperms yet. I am going to make a PR, but now is the last chance to change the variable name if anyone has good suggestions.

The x86_64 check is failing because I forgot to fix permissions for one path in vscode, I will fix it on the next push.

ericonr commented on common/hooks/post-install/14-fix-permissions.sh:
```diff
@@ -0,0 +1,30 @@
+}
+
+hook() {
+	[ -n "$nofixperms" ] && return 0
```
So `nofixperms` also implies not checking perms at all, correct? I wonder if we should have `nocheckperms` for this step, and then `nofixperms` below?

ericonr replied:
I'm not sure this level of granularity is actually necessary, just a suggestion.

paper42 replied:
good idea, I modified it a bit to make it a bit more intuitive in my opinion:

* `nocheckperms` If set, xbps-src will not fail on common permission errors (world writable files, etc.)
* `nofixperms` If set, xbps-src will not fix common permission errors (executable manpages, etc.)

So the variables are entirely independent now, if nocheckperms is set, only fix will run, if both are set, the hook will not do anything, etc.

ericonr reviewed (commented):
This is missing docs in Manual.md.
